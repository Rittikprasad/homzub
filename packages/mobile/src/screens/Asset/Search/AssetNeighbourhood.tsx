/* eslint-disable react/no-unused-state */
import React from 'react';
import { StyleSheet, View } from 'react-native';
import { connect } from 'react-redux';
import MapView, { Marker, PROVIDER_GOOGLE } from 'react-native-maps';
import { WithTranslation, withTranslation } from 'react-i18next';
import { AlertHelper } from '@homzhub/common/src/utils/AlertHelper';
import { GooglePlacesService } from '@homzhub/common/src/services/GooglePlaces/GooglePlacesService';
import { theme } from '@homzhub/common/src/styles/theme';
import Icon, { icons } from '@homzhub/common/src/assets/icon';
import { AssetSelectors } from '@homzhub/common/src/modules/asset/selectors';
import { UserSelector } from '@homzhub/common/src/modules/user/selectors';
import { CustomMarker } from '@homzhub/common/src/components/atoms/CustomMarker';
import { SelectionPicker } from '@homzhub/common/src/components/atoms/SelectionPicker';
import { BottomSheet } from '@homzhub/common/src/components/molecules/BottomSheet';
import ExploreSections from '@homzhub/mobile/src/components/molecules/ExploreSections';
import { Header } from '@homzhub/mobile/src/components/molecules/Header';
import { Asset } from '@homzhub/common/src/domain/models/Asset';
import { NeighborhoodTabs, PLACES_DATA } from '@homzhub/common/src/constants/AssetNeighbourhood';
import { LocaleConstants } from '@homzhub/common/src/services/Localization/constants';
import { MetricSystems } from '@homzhub/common/src/domain/models/UserPreferences';
import { PlaceTypes } from '@homzhub/common/src/services/GooglePlaces/constants';
import { IState } from '@homzhub/common/src/modules/interfaces';
import { SearchStackParamList } from '@homzhub/mobile/src/navigation/SearchStack';
import { PointOfInterest } from '@homzhub/common/src/services/GooglePlaces/interfaces';
import { NavigationScreenProps, ScreensKeys } from '@homzhub/mobile/src/navigation/interfaces';

const EXPANDED = theme.viewport.height * 0.45;

// INTERFACES & TYPES
interface IStateProps {
  asset: Asset | null;
  metricSystem: MetricSystems;
}
interface IOwnState {
  selectedTab: NeighborhoodTabs;
  selectedPlaceType: PlaceTypes;
  isBottomSheetVisible: boolean;
  pointsOfInterest: PointOfInterest[];
  selectedPlaceId: string;
  isApiActive: boolean;
}

type libraryProps = WithTranslation & NavigationScreenProps<SearchStackParamList, ScreensKeys.AssetNeighbourhood>;
type Props = IStateProps & libraryProps;

export class AssetNeighbourhood extends React.Component<Props, IOwnState> {
  private mapRef: MapView | null = null;

  public state = {
    isBottomSheetVisible: false,
    selectedTab: NeighborhoodTabs.Nearby,
    selectedPlaceType: Object.values(PLACES_DATA[NeighborhoodTabs.Nearby])[0].key,
    pointsOfInterest: [],
    selectedPlaceId: '',
    isApiActive: false,
  };

  public componentDidMount = async (): Promise<void> => {
    await this.fetchPOIs();
    this.setState({ isBottomSheetVisible: true });
  };

  public shouldComponentUpdate = (nextProps: Props, nextState: IOwnState): boolean => {
    const { isApiActive, selectedPlaceType, selectedTab } = nextState;
    const { selectedPlaceType: oldType, selectedTab: oldTab } = this.state;
    return !(isApiActive || selectedTab !== oldTab || selectedPlaceType !== oldType);
  };

  public render = (): React.ReactNode => {
    const { asset, t, metricSystem } = this.props;
    const { selectedTab, selectedPlaceType, isBottomSheetVisible, pointsOfInterest, selectedPlaceId } = this.state;

    if (!asset) return null;
    const {
      assetLocation: { longitude, latitude },
    } = asset;
    return (
      <View style={styles.flexOne}>
        <Header type="secondary" icon={icons.leftArrow} onIconPress={this.onBackPress} title={asset.projectName} />
        <MapView
          ref={(mapRef: MapView): void => {
            this.mapRef = mapRef;
          }}
          provider={PROVIDER_GOOGLE}
          style={styles.flexOne}
          initialRegion={{
            latitude,
            longitude,
            latitudeDelta: 0.1,
            longitudeDelta: 0.1,
          }}
        >
          <Marker coordinate={{ latitude, longitude }}>
            <CustomMarker selected />
          </Marker>
          {this.renderMarkers()}
        </MapView>
        <View style={styles.floatingTab}>
          <SelectionPicker
            data={[
              { title: t('nearby'), value: NeighborhoodTabs.Nearby },
              { title: t('commute'), value: NeighborhoodTabs.Commute },
            ]}
            selectedItem={[selectedTab]}
            onValueChange={this.onTabChange}
          />
        </View>
        <BottomSheet
          visible={isBottomSheetVisible}
          renderHeader={false}
          sheetHeight={EXPANDED}
          onCloseSheet={this.onBottomSheetClose}
        >
          <ExploreSections
            placeTypes={Object.values(PLACES_DATA[selectedTab])}
            // @ts-ignore
            selectedPlaceType={PLACES_DATA[selectedTab][selectedPlaceType]}
            onPlaceTypePress={this.onPlaceTypeChange}
            selectedPoiId={selectedPlaceId}
            pointsOfInterest={pointsOfInterest}
            onPoiPress={this.onPoiPress}
            metricSystem={metricSystem}
          />
        </BottomSheet>
      </View>
    );
  };

  private renderMarkers = (): React.ReactNode => {
    const { pointsOfInterest, selectedPlaceId, selectedTab, selectedPlaceType } = this.state;
    // @ts-ignore
    const placeType = PLACES_DATA[selectedTab][selectedPlaceType];

    return pointsOfInterest.map((point: PointOfInterest) => (
      <Marker key={point.placeId} coordinate={{ latitude: point.latitude, longitude: point.longitude }}>
        <Icon
          name={placeType.mapMarker}
          color={selectedPlaceId === point.placeId ? theme.colors.active : theme.colors.darkTint5}
          size={24}
        />
      </Marker>
    ));
  };

  private onTabChange = (selectedTab: NeighborhoodTabs): void => {
    const { selectedTab: oldTab, selectedPlaceType } = this.state;

    let placeTypeToUpdate = selectedPlaceType;
    if (oldTab !== selectedTab) {
      placeTypeToUpdate = Object.values(PLACES_DATA[selectedTab])[0].key;
    }

    this.setState({ selectedTab, selectedPlaceType: placeTypeToUpdate, isBottomSheetVisible: true }, () => {
      this.fetchPOIs().then();
    });
  };

  private onPlaceTypeChange = (selectedPlaceType: PlaceTypes): void => {
    this.setState({ selectedPlaceType }, () => {
      this.fetchPOIs().then();
    });
  };

  private onBottomSheetClose = (): void => {
    this.setState({ isBottomSheetVisible: false });
  };

  private onPoiPress = (poi: PointOfInterest): void => {
    const { latitude, longitude, placeId } = poi;
    this.setState({ selectedPlaceId: placeId });
    this.mapRef?.animateCamera({
      center: {
        latitude,
        longitude,
      },
    });
  };

  private onBackPress = (): void => {
    const {
      navigation: { goBack },
    } = this.props;
    goBack();
  };

  private fetchPOIs = async (): Promise<void> => {
    const { selectedPlaceType } = this.state;
    const { asset, metricSystem } = this.props;

    if (!asset) return;
    this.setState({ isApiActive: true });
    const { assetLocation } = asset;

    try {
      const pointsOfInterest = await GooglePlacesService.getPOIs(
        assetLocation,
        selectedPlaceType,
        undefined,
        metricSystem
      );

      this.setState({ pointsOfInterest, isApiActive: false });
    } catch (e) {
      this.setState({ isBottomSheetVisible: false, isApiActive: false });
      AlertHelper.error({ message: e.message });
    }
  };
}

const styles = StyleSheet.create({
  flexOne: {
    flex: 1,
  },
  floatingTab: {
    position: 'absolute',
    alignItems: 'center',
    left: 0,
    right: 0,
    marginHorizontal: 12,
    top: 120,
  },
});

const mapStateToProps = (state: IState): IStateProps => {
  const { getAsset } = AssetSelectors;
  const { getMetricSystem } = UserSelector;

  return {
    asset: getAsset(state),
    metricSystem: getMetricSystem(state),
  };
};

export default connect(
  mapStateToProps,
  null
)(withTranslation(LocaleConstants.namespacesKey.assetDescription)(AssetNeighbourhood));
