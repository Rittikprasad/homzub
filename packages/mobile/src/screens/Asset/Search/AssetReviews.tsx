import React, { memo, useState, useEffect, useRef } from 'react';
import { StyleSheet, View } from 'react-native';
import { useSelector } from 'react-redux';
import { useTranslation } from 'react-i18next';
import { AlertHelper } from '@homzhub/common/src/utils/AlertHelper';
import { AssetRepository } from '@homzhub/common/src/domain/repositories/AssetRepository';
import { theme } from '@homzhub/common/src/styles/theme';
import { AssetSelectors } from '@homzhub/common/src/modules/asset/selectors';
import { Dropdown } from '@homzhub/common/src/components/atoms/Dropdown';
import { Label } from '@homzhub/common/src/components/atoms/Text';
import { AssetReviewCard } from '@homzhub/mobile/src/components/molecules/AssetReviewCard';
import { AssetReviewsSummary } from '@homzhub/mobile/src/components/molecules/AssetReviewsSummary';
import { Screen } from '@homzhub/mobile/src/components/HOC/Screen';
import { AssetReview } from '@homzhub/common/src/domain/models/AssetReview';
import { SearchStackParamList } from '@homzhub/mobile/src/navigation/SearchStack';
import { NavigationScreenProps, ScreensKeys } from '@homzhub/mobile/src/navigation/interfaces';
import { LocaleConstants } from '@homzhub/common/src/services/Localization/constants';

const SLIDER_WIDTH = theme.viewport.width - theme.layout.screenPadding * 2;
enum SortType {
  MostRecent = 'MostRecent',
  Positive = 'Positive',
  Critical = 'Critical',
}
type Props = NavigationScreenProps<SearchStackParamList, ScreensKeys.AssetReviews>;

const AssetReviews = (props: Props): React.ReactElement => {
  const {
    navigation,
    route: { params },
  } = props;
  const reviewSummary = useSelector(AssetSelectors.getAssetReviews);
  const { t } = useTranslation(LocaleConstants.namespacesKey.assetDescription);

  const [reviews, setReviews] = useState<AssetReview[]>([]);
  const [sortType, setSortType] = useState<SortType>(SortType.MostRecent);
  const [loading, setLoading] = useState(true);
  const dropDownOptions = useRef([
    { value: SortType.MostRecent, label: t('mostRecent') },
    { value: SortType.Positive, label: t('positive') },
    { value: SortType.Critical, label: t('critical') },
  ]);

  useEffect(() => {
    try {
      AssetRepository.getListingReviews(params).then((response) => {
        setReviews(response.sort((a, b) => new Date(b.modifiedAt).getTime() - new Date(a.modifiedAt).getTime()));
      });
    } catch (err) {
      AlertHelper.error({ message: t('common:genericErrorMessage') });
    } finally {
      setLoading(false);
    }
  }, []);

  useEffect(() => {
    if (reviews.length <= 0) {
      return;
    }

    let sorted: AssetReview[] = [];

    switch (sortType) {
      case SortType.Positive:
        sorted = [...reviews].sort((a, b) => b.rating - a.rating);
        break;
      case SortType.Critical:
        sorted = [...reviews].sort((a, b) => a.rating - b.rating);
        break;
      default:
        sorted = [...reviews].sort((a, b) => new Date(b.modifiedAt).getTime() - new Date(a.modifiedAt).getTime());
        break;
    }

    setReviews(sorted);
  }, [sortType]);

  return (
    <Screen
      backgroundColor={theme.colors.white}
      headerProps={{
        type: 'secondary',
        title: t('customerReviews'),
        onIconPress: navigation.goBack,
      }}
      contentContainerStyle={styles.screen}
      isLoading={loading}
    >
      {reviewSummary && (
        <AssetReviewsSummary reviewSummary={reviewSummary} containerStyle={styles.padding} sliderWidth={SLIDER_WIDTH} />
      )}
      <View style={styles.heading}>
        <Label type="large" textType="semiBold" style={styles.reviewHeading}>
          {t('property:totalReviews', { total: reviews.length })}
        </Label>
        <Dropdown
          data={dropDownOptions.current}
          value={sortType}
          onDonePress={setSortType}
          listHeight={theme.viewport.height / 2}
          testID="drpTimeRange"
          isOutline
          containerStyle={styles.dropdownStyle}
          fontSize="regular"
        />
      </View>
      {reviews.map((review: AssetReview) => (
        <AssetReviewCard key={review.id} review={review} hideButtons />
      ))}
    </Screen>
  );
};

const memoizedComponent = memo(AssetReviews);
export { memoizedComponent as AssetReviews };

const styles = StyleSheet.create({
  screen: {
    paddingHorizontal: 0,
  },
  padding: {
    marginHorizontal: 16,
  },
  heading: {
    marginHorizontal: 16,
    marginVertical: 20,
    flexDirection: 'row',
    justifyContent: 'space-between',
  },
  reviewHeading: {
    color: theme.colors.darkTint3,
  },
  dropdownStyle: {
    width: 132,
  },
});
