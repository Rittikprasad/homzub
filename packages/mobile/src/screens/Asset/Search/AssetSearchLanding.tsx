import React from 'react';
import { View, StyleSheet, StatusBar, PickerItemProps, ScrollView } from 'react-native';
import { bindActionCreators, Dispatch } from 'redux';
import { connect } from 'react-redux';
import { withTranslation, WithTranslation } from 'react-i18next';
import { GeolocationResponse } from '@homzhub/common/src/services/Geolocation/interfaces';
import { debounce } from 'lodash';
import { AlertHelper } from '@homzhub/common/src/utils/AlertHelper';
import { PlatformUtils } from '@homzhub/common/src/utils/PlatformUtils';
import { LocaleConstants } from '@homzhub/common/src/services/Localization/constants';
import { GooglePlaceData, GooglePlaceDetail } from '@homzhub/common/src/services/GooglePlaces/interfaces';
import { GooglePlacesService } from '@homzhub/common/src/services/GooglePlaces/GooglePlacesService';
import { IState } from '@homzhub/common/src/modules/interfaces';
import { CommonSelectors } from '@homzhub/common/src/modules/common/selectors';
import { SearchSelector } from '@homzhub/common/src/modules/search/selectors';
import { SearchActions } from '@homzhub/common/src/modules/search/actions';
import { theme } from '@homzhub/common/src/styles/theme';
import Icon, { icons } from '@homzhub/common/src/assets/icon';
import { Button } from '@homzhub/common/src/components/atoms/Button';
import { Loader } from '@homzhub/common/src/components/atoms/Loader';
import { SelectionPicker } from '@homzhub/common/src/components/atoms/SelectionPicker';
import { Text } from '@homzhub/common/src/components/atoms/Text';
import { WithShadowView } from '@homzhub/common/src/components/atoms/WithShadowView';
import { CurrentLocation } from '@homzhub/mobile/src/components';
import { Range } from '@homzhub/common/src/components/molecules/Range';
import SearchResults from '@homzhub/mobile/src/components/molecules/SearchResults';
import GoogleSearchBar from '@homzhub/mobile/src/components/molecules/GoogleSearchBar';
import { NavigationScreenProps, ScreensKeys } from '@homzhub/mobile/src/navigation/interfaces';
import { WrapperSearchStackParamList } from '@homzhub/mobile/src/navigation/WrapperSearchStack';
import { Country } from '@homzhub/common/src/domain/models/Country';
import { FilterDetail } from '@homzhub/common/src/domain/models/FilterDetail';
import { IFilter, ITransactionRange } from '@homzhub/common/src/domain/models/Search';

interface IStateProps {
  filterData: FilterDetail | null;
  filters: IFilter;
  currencyData: PickerItemProps[];
  priceRange: ITransactionRange;
  countryList: Country[];
  isLoading: boolean;
}

interface IDispatchProps {
  getFilterDetails: (payload: any) => void;
  setFilter: (payload: any) => void;
  getProperties: () => void;
  setInitialState: () => void;
}

type libraryProps = NavigationScreenProps<WrapperSearchStackParamList, ScreensKeys.PropertySearchLanding>;
type Props = IStateProps & IDispatchProps & libraryProps & WithTranslation;

interface ILandingState {
  searchString: string;
  isSearchBarFocused: boolean;
  suggestions: GooglePlaceData[];
  selectedPropertyType: number;
  selectedLookingType: number;
  maxPriceRange: number;
  minPriceRange: number;
  isLocationSelected: boolean;
}

export class AssetSearchLanding extends React.PureComponent<Props, ILandingState> {
  public state = {
    searchString: '',
    isSearchBarFocused: false,
    suggestions: [],
    selectedPropertyType: 0,
    selectedLookingType: 0,
    maxPriceRange: 0,
    minPriceRange: 0,
    isLocationSelected: false,
  };

  public componentDidMount = (): void => {
    const { getFilterDetails, filters } = this.props;
    const { asset_group, asset_transaction_type, min_price, max_price } = filters;
    this.setState({
      selectedPropertyType: asset_group ?? 0,
      selectedLookingType: asset_transaction_type ?? 0,
      minPriceRange: min_price ?? 0,
      maxPriceRange: max_price ?? 0,
    });
    getFilterDetails({ asset_group: filters.asset_group });
  };

  public componentDidUpdate = (prevProps: Props): void => {
    const {
      filters: { asset_group, max_price, min_price, search_latitude, asset_transaction_type },
      getFilterDetails,
    } = this.props;

    if (prevProps.filters.asset_group !== asset_group || prevProps.filters.search_latitude !== search_latitude) {
      getFilterDetails({ asset_group });
    }

    // eslint-disable-next-line react/no-did-update-set-state
    this.setState({
      selectedPropertyType: asset_group ?? 0,
      selectedLookingType: asset_transaction_type ?? 0,
      minPriceRange: min_price ?? 0,
      maxPriceRange: max_price ?? 0,
    });
  };

  public componentWillUnmount = (): void => {
    const { setInitialState } = this.props;
    setInitialState();
  };

  public render(): React.ReactElement {
    const { isSearchBarFocused, isLocationSelected } = this.state;
    const { t, filterData, isLoading } = this.props;
    return (
      <>
        <View style={styles.statusBar}>
          <StatusBar translucent backgroundColor={theme.colors.background} barStyle="dark-content" />
        </View>
        <View style={styles.screen}>
          {this.renderHeader()}
          {isSearchBarFocused && this.renderSearchResults()}
          {!isSearchBarFocused && filterData && this.renderContent(filterData)}
        </View>
        <WithShadowView outerViewStyle={styles.shadowView}>
          <Button
            type="primary"
            title={t('showProperties')}
            disabled={!isLocationSelected}
            containerStyle={styles.buttonStyle}
            onPress={this.onShowProperties}
          />
        </WithShadowView>
        <Loader visible={isLoading} />
      </>
    );
  }

  private renderContent = (filterData: FilterDetail): React.ReactElement => {
    const { selectedPropertyType, selectedLookingType, minPriceRange, maxPriceRange } = this.state;
    const {
      t,
      currencyData,
      priceRange,
      filters: { currency_code },
      countryList,
    } = this.props;
    const {
      currency,
      assetGroupList,
      filters: { transactionType },
    } = filterData;
    let currencySymbol = '';

    const assetGroup = assetGroupList.map((item, index) => {
      return { title: item.title, value: item.id };
    });

    // TODO: Handle Multiple currency
    const country = countryList.find((item) => item.currencies[0].currencyCode === currency_code);

    currencySymbol = country?.currencies[0].currencySymbol ?? currency[0].currencySymbol;

    const assetTransaction = transactionType.map((item, index) => {
      return { title: item.title, value: index };
    });

    return (
      <ScrollView
        style={styles.content}
        contentContainerStyle={styles.contentContainer}
        showsVerticalScrollIndicator={false}
      >
        <Text type="small" textType="semiBold" style={styles.label}>
          {t('propertyType')}
        </Text>
        <SelectionPicker
          data={assetGroup}
          selectedItem={[selectedPropertyType]}
          onValueChange={this.onChangeProperty}
        />
        <Text type="small" textType="semiBold" style={styles.label}>
          {t('lookingFor')}
        </Text>
        <SelectionPicker
          data={assetTransaction}
          selectedItem={[selectedLookingType]}
          onValueChange={this.onChangeFlow}
        />
        <Range
          //  @ts-ignore
          dropdownData={currencyData}
          isPriceRange
          selectedUnit={currency_code}
          currencySymbol={currencySymbol}
          onChangeSlide={this.updateFilter}
          range={priceRange}
          minChangedValue={minPriceRange}
          maxChangedValue={maxPriceRange}
          containerStyle={styles.priceRange}
        />
      </ScrollView>
    );
  };

  private renderHeader = (): React.ReactNode => {
    const { t } = this.props;
    const { searchString } = this.state;

    return (
      <View style={styles.header}>
        <Icon
          size={24}
          name={icons.leftArrow}
          color={theme.colors.darkTint2}
          style={styles.backIconStyle}
          onPress={this.onGoBack}
        />
        <Text type="regular">{t('findingRightProperty')}</Text>
        <Text type="regular" textType="bold" style={styles.madeEasy}>
          {t('madeEasy')}
        </Text>
        <GoogleSearchBar
          placeholder={t('enterLocation')}
          updateValue={this.onSearchStringUpdate}
          value={searchString}
          containerStyle={styles.searchBarContainer}
          cancelButtonStyle={styles.cancelButtonStyle}
          cancelTextStyle={styles.cancelTextStyle}
          onFocusChange={this.onSearchBarFocusChange}
        />
      </View>
    );
  };

  private renderSearchResults = (): React.ReactNode => {
    const { suggestions, searchString } = this.state;

    return (
      <>
        <CurrentLocation onGetCurrentPositionSuccess={this.onGetCurrentPositionSuccess} />
        {suggestions.length > 0 && searchString.length > 0 && (
          <SearchResults
            results={suggestions}
            onResultPress={this.onSuggestionPress}
            listTitleStyle={styles.resultListContainer}
          />
        )}
      </>
    );
  };

  private onSearchBarFocusChange = (isSearchBarFocused: boolean): void => {
    this.setState({ isSearchBarFocused });
  };

  private onGoBack = (): void => {
    const { navigation } = this.props;
    navigation.goBack();
  };

  private onSearchStringUpdate = (searchString: string): void => {
    this.setState({ searchString }, this.getAutocompleteSuggestions);
  };

  private onSuggestionPress = (place: GooglePlaceData): void => {
    const { setFilter } = this.props;
    GooglePlacesService.getPlaceDetail(place.place_id)
      .then((placeDetail: GooglePlaceDetail) => {
        this.setPropertiesCurrency(placeDetail);
        setFilter({
          search_address: place.description,
          search_latitude: placeDetail.geometry.location.lat,
          search_longitude: placeDetail.geometry.location.lng,
        });
        this.setState({ isLocationSelected: true, isSearchBarFocused: false, searchString: place.description });
      })
      .catch(this.displayError);
  };

  private onGetCurrentPositionSuccess = (data: GeolocationResponse): void => {
    const { setFilter } = this.props;
    const {
      coords: { latitude, longitude },
    } = data;
    GooglePlacesService.getLocationData({ lng: longitude, lat: latitude }).then((locData) => {
      const { formatted_address } = locData;
      const { primaryAddress, secondaryAddress } = GooglePlacesService.getSplitAddress(formatted_address);
      setFilter({
        search_address: `${primaryAddress} ${secondaryAddress}`,
        search_latitude: latitude,
        search_longitude: longitude,
      });
      this.setState({
        isLocationSelected: true,
        isSearchBarFocused: false,
        searchString: `${primaryAddress} ${secondaryAddress}`,
      });
    });
  };

  private onShowProperties = (): void => {
    const { navigation } = this.props;
    navigation.navigate(ScreensKeys.Search);
  };

  // eslint-disable-next-line react/sort-comp
  private getAutocompleteSuggestions = debounce((): void => {
    const { searchString } = this.state;

    GooglePlacesService.autoComplete(searchString)
      .then((suggestions: GooglePlaceData[]) => {
        this.setState({ suggestions });
      })
      .catch((e: Error): void => {
        AlertHelper.error({ message: e.message });
      });
  }, 500);

  private onChangeProperty = (value: number): void => {
    const { setFilter } = this.props;
    setFilter({ asset_group: value, min_price: -1, max_price: -1, asset_transaction_type: 0 });
  };

  private onChangeFlow = (value: number): void => {
    const { setFilter } = this.props;
    setFilter({ asset_transaction_type: value, min_price: -1, max_price: -1 });
  };

  private updateFilter = (type: string, value: number | number[]): void => {
    const { setFilter } = this.props;
    setFilter({ [type]: value });
  };

  private displayError = (e: Error): void => {
    AlertHelper.error({ message: e.message });
  };

  private setPropertiesCurrency = (placeDetail: GooglePlaceDetail): void => {
    const { countryList, setFilter } = this.props;
    const placeCountry = placeDetail.address_components.find((address) => address.types.includes('country'));
    const country = countryList.find((item) => item.iso2Code === placeCountry?.short_name);
    setFilter({
      currency_code: country?.currencies[0].currencyCode,
    });
  };
}

export const mapStateToProps = (state: IState): IStateProps => {
  return {
    filterData: SearchSelector.getFilterDetail(state),
    filters: SearchSelector.getFilters(state),
    currencyData: SearchSelector.getCurrencyData(state),
    priceRange: SearchSelector.getPriceRange(state),
    isLoading: SearchSelector.getLoadingState(state),
    countryList: CommonSelectors.getCountryList(state),
  };
};

export const mapDispatchToProps = (dispatch: Dispatch): IDispatchProps => {
  const { getFilterDetails, setFilter, getProperties, setInitialState } = SearchActions;
  return bindActionCreators(
    {
      getFilterDetails,
      setFilter,
      getProperties,
      setInitialState,
    },
    dispatch
  );
};

export default connect<IStateProps, IDispatchProps, WithTranslation, IState>(
  mapStateToProps,
  mapDispatchToProps
)(withTranslation(LocaleConstants.namespacesKey.propertySearch)(AssetSearchLanding));

const styles = StyleSheet.create({
  screen: {
    flex: 1,
    backgroundColor: theme.colors.white,
  },
  statusBar: {
    marginTop: 16,
    height: PlatformUtils.isIOS() ? 30 : StatusBar.currentHeight,
    backgroundColor: theme.colors.background,
  },
  header: {
    paddingHorizontal: theme.layout.screenPadding,
    paddingVertical: 8,
    backgroundColor: theme.colors.background,
  },
  searchBarContainer: {
    backgroundColor: theme.colors.background,
    paddingHorizontal: 0,
    paddingBottom: 0,
    paddingTop: 24,
  },
  cancelButtonStyle: {
    backgroundColor: theme.colors.background,
  },
  cancelTextStyle: {
    color: theme.colors.primaryColor,
  },
  resultListContainer: {
    backgroundColor: theme.colors.white,
  },
  madeEasy: {
    marginTop: 6,
  },
  shadowView: {
    paddingTop: 10,
    paddingBottom: 0,
  },
  buttonStyle: {
    flex: 0,
    margin: 16,
  },
  content: {
    backgroundColor: theme.colors.white,
    flex: 1,
  },
  contentContainer: {
    padding: theme.layout.screenPadding,
  },
  label: {
    color: theme.colors.darkTint4,
    marginVertical: 20,
  },
  priceRange: {
    marginVertical: 30,
  },
  backIconStyle: {
    paddingVertical: 16,
    marginBottom: 10,
  },
});
