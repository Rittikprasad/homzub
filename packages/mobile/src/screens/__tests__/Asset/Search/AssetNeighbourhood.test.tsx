import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import toJson from 'enzyme-to-json';
import { AssetNeighbourhood } from '@homzhub/mobile/src/screens/Asset/Search/AssetNeighbourhood';
import { mockAsset } from '@homzhub/common/src/mocks/AssetDescription';

describe('AssetNeighbourhood', () => {
  let component: ShallowWrapper;
  let props: any;
  const mock = jest.fn();

  beforeEach(() => {
    props = {
      asset: mockAsset,
      navigation: {
        goBack: mock,
      },
    };
  });

  it('should render snapshot', () => {
    component = shallow(<AssetNeighbourhood {...props} t={(key: string): string => key} />);
    expect(toJson(component)).toMatchSnapshot();
  });
});
