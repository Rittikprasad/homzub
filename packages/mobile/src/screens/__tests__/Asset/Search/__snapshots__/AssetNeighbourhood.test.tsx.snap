// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AssetNeighbourhood should render snapshot 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <Memo(Header)
    icon="left-arrow"
    onIconPress={[Function]}
    type="secondary"
  />
  <MapView
    initialRegion={
      Object {
        "latitude": 25.619,
        "latitudeDelta": 0.1,
        "longitude": 85.123,
        "longitudeDelta": 0.1,
      }
    }
    provider="google"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <MapMarker
      coordinate={
        Object {
          "latitude": 25.619,
          "longitude": 85.123,
        }
      }
      stopPropagation={false}
    >
      <Memo(CustomMarker)
        selected={true}
      />
    </MapMarker>
  </MapView>
  <View
    style={
      Object {
        "alignItems": "center",
        "left": 0,
        "marginHorizontal": 12,
        "position": "absolute",
        "right": 0,
        "top": 120,
      }
    }
  >
    <SelectionPicker
      data={
        Array [
          Object {
            "title": "nearby",
            "value": 0,
          },
          Object {
            "title": "commute",
            "value": 1,
          },
        ]
      }
      onValueChange={[Function]}
      selectedItem={
        Array [
          0,
        ]
      }
    />
  </View>
  <BottomSheet
    onCloseSheet={[Function]}
    renderHeader={false}
    sheetHeight={600.3000000000001}
    visible={false}
  >
    <withI18nextTranslation(ExploreSections)
      onPlaceTypePress={[Function]}
      onPoiPress={[Function]}
      placeTypes={
        Array [
          Object {
            "icon": "school",
            "key": "school",
            "label": "schools",
            "mapMarker": "school-marker",
          },
          Object {
            "icon": "mall",
            "key": "shopping_mall",
            "label": "malls",
            "mapMarker": "mall-marker",
          },
          Object {
            "icon": "restaurant",
            "key": "restaurant",
            "label": "restaurants",
            "mapMarker": "restaurant-marker",
          },
          Object {
            "icon": "fuel",
            "key": "gas_station",
            "label": "fuelStations",
            "mapMarker": "fuel-marker",
          },
          Object {
            "icon": "hospital",
            "key": "hospital",
            "label": "hospitals",
            "mapMarker": "hospital-marker",
          },
          Object {
            "icon": "grocery",
            "key": "supermarket",
            "label": "groceryStores",
            "mapMarker": "grocery-marker",
          },
          Object {
            "icon": "park",
            "key": "park",
            "label": "parks",
            "mapMarker": "park-marker",
          },
          Object {
            "icon": "cinema",
            "key": "movie_theater",
            "label": "filmTheaters",
            "mapMarker": "cinema-marker",
          },
          Object {
            "icon": "bank",
            "key": "atm",
            "label": "atms",
            "mapMarker": "bank-marker",
          },
          Object {
            "icon": "chemist",
            "key": "pharmacy",
            "label": "pharmacies",
            "mapMarker": "chemist-marker",
          },
          Object {
            "icon": "lodge",
            "key": "lodging",
            "label": "hotels",
            "mapMarker": "lodge-marker",
          },
        ]
      }
      pointsOfInterest={Array []}
      selectedPlaceType={
        Object {
          "icon": "school",
          "key": "school",
          "label": "schools",
          "mapMarker": "school-marker",
        }
      }
      selectedPoiId=""
    />
  </BottomSheet>
</View>
`;
